### PR TITLE
Created azure-storage unit tests (excluding TransferManager). Modified public API.

### PR DESCRIPTION
**Pull request by vcolin7**
Thanks for taking the time to contribute. Please take a moment to fill out the information below.

Partially addresses #189.

Changes Proposed in this pull request:
- Created unit tests for azure-storage interceptors.
- Created unit tests for StorageBlobClient.
- Polished the StorageBlobClient public API in to be more consistent.

### CODE REVIEW

File: sdk/storage/azure-storage-blob/src/main/java/com/azure/android/storage/blob/StorageBlobServiceImpl.java [hunk 1]
```diff
@@ -921,15 +1015,19 @@ public void onFailure(@NonNull Call<ResponseBody> call, @NonNull Throwable t) {
...
 
                     return new CallAndOptionalResult<>(call, result);
-                } else { // Error response
-                    String strContent = readAsString(response.body());
+                } else {
+                    ResponseBody responseBody = response.body();
+                    String message = responseBody == null ? "no message" : "message: " + readAsString(responseBody);
 
-                    throw new BlobStorageException(strContent, response.raw());
+                    throw new BlobStorageException("Response failed with code: " + response.code() + " and " + message,
```
Comment (anuchandy) [inline sdk/storage/azure-storage-blob/src/main/java/com/azure/android/storage/blob/StorageBlobServiceImpl.java:1022]: Refer above comment i.e. I would suggest keeping the message as it is. Also the HTTP response.code() is often 200 with the actual error-code exists in the response-header, so this extra text additions may not add much value but make it bit difficult to parse like I mentioned above.

File: sdk/storage/azure-storage-blob/src/main/java/com/azure/android/storage/blob/interceptor/SasTokenCredentialInterceptor.java [hunk 2]
```diff
@@ -28,17 +28,20 @@ public SasTokenCredentialInterceptor(SasTokenCredential credential) {
...
             : sasToken;
 
+        // SAS token is already encoded so its safe to append it to the encoded query from source request.
+        String encodedQuery = requestURL.encodedQuery();
+
+        if (isNullOrEmpty(encodedQuery)) {
+            encodedQuery = sasToken;
+        } else {
+            encodedQuery += "&" + sasToken;
+        }
```
Comment (anuchandy) [inline sdk/storage/azure-storage-blob/src/main/java/com/azure/android/storage/blob/interceptor/SasTokenCredentialInterceptor.java:43]: good catch and fix 👍

File: sdk/storage/azure-storage-blob/src/main/java/com/azure/android/storage/blob/StorageBlobClient.java [hunk 3]
```diff
@@ -392,28 +399,28 @@ public ServiceCall downloadWithHeaders(String containerName,
             callback);
     }
 
-    public void stageBlock(String containerName,
+    public BlockBlobsStageBlockResponse stageBlock(String containerName,
```
Comment (anuchandy) [inline sdk/storage/azure-storage-blob/src/main/java/com/azure/android/storage/blob/StorageBlobClient.java:402]: The methods in this category are those with "simplest" signature, which should not return any `Response` type but instead returns the actual value in the response, in this case it should be `void`.

File: sdk/storage/azure-storage-blob/src/main/java/com/azure/android/storage/blob/StorageBlobClient.java [hunk 4]
```diff
@@ -392,28 +399,28 @@ public ServiceCall downloadWithHeaders(String containerName,
...
     }
 
     public ServiceCall stageBlock(String containerName,
                                   String blobName,
                                   String base64BlockId,
-                                  byte[] body,
+                                  byte[] blockContent,
                                   byte[] contentMd5,
-                                  Callback<Void> callback) {
+                                  Callback<BlockBlobsStageBlockResponse> callback) {
```
Comment (anuchandy) [inline sdk/storage/azure-storage-blob/src/main/java/com/azure/android/storage/blob/StorageBlobClient.java:419]: ref the previous comment, this is the async variant with "simplest" signature, and should be producing value in the Response (i.e. void) not the Response.

File: sdk/storage/azure-storage-blob/src/main/java/com/azure/android/storage/blob/StorageBlobClient.java (hunk 3 above)
Comment (vcolin7) [reply sdk/storage/azure-storage-blob/src/main/java/com/azure/android/storage/blob/StorageBlobClient.java:402]: You are right, I must have missed this for `stageBlock`.

File: sdk/storage/azure-storage-blob/src/main/java/com/azure/android/storage/blob/StorageBlobClient.java (hunk 4 above)
Comment (vcolin7) [reply sdk/storage/azure-storage-blob/src/main/java/com/azure/android/storage/blob/StorageBlobClient.java:419]: So we would make this accept a `Callback<Void>` instead, right?

Comment (anuchandy) [reply sdk/storage/azure-storage-blob/src/main/java/com/azure/android/storage/blob/StorageBlobClient.java]: right, I see you already made this change, sorry for delay.

File: sdk/core/azure-core/src/main/java/com/azure/android/core/internal/util/test/EnqueueMockResponse.java [hunk 5]
```diff
@@ -1,4 +1,4 @@
-package com.azure.android.core.http.interceptor;
+package com.azure.android.core.internal.util.test;
```
Comment (anuchandy) [inline sdk/core/azure-core/src/main/java/com/azure/android/core/internal/util/test/EnqueueMockResponse.java:1]: I guess there are two things in the `test` project:

1. Unit tests for types in the `core` project.
2. Test types that we want to share with other "dependent" projects (e.g., storage).

For the second category, does it make sense to keeping these shared types rooted at `com.azure.android.core.test,` hence consistent with the established namespace in java [sdk](https://github.com/Azure/azure-sdk-for-java/tree/master/sdk/core/azure-core-test/src/main/java/com/azure/core/test). Its a shippable package in java SDK, we haven't there yet for android but agreeing on the namespace will be a good start. Thoughts?

Comment (anuchandy) [reply sdk/core/azure-core/src/main/java/com/azure/android/core/internal/util/test/EnqueueMockResponse.java:1]: Between why did we move "shared test" types from "src/test" to "src/main"?

Comment (anuchandy) [reply sdk/core/azure-core/src/main/java/com/azure/android/core/internal/util/test/EnqueueMockResponse.java:1]: If discoverability of types are the reason for "src/test" -> "src/main" then worth checking this: https://stackoverflow.com/questions/5644011/multi-project-test-dependencies-with-gradle/60138176#60138176

**Comment by anuchandy** [comment]
@vcolin7 should we merge this?

**Comment by vcolin7** [comment]
@anuchandy Let's merge this for now and we will revisit the strategy to include core test classes at a later point.

**Pull request merged**